### PR TITLE
Update coarsening

### PR DESCRIPTION
**Pull request by alessiofumagalli**
- update the coarsening procedure 
- introduce a new way of gridding (2d) with gmsh
- new examples for VEM with internal boundaries
- NOTE: some tests are not active for python 2.7 and 3.5

### CODE REVIEW

File: examples/example9/continuous.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    left = seeds[g.cell_centers[0, seeds] < 0.5]
+    right = seeds[g.cell_centers[0, seeds] > 0.5]
+
+    partition[partition == partition[left[1]]] = partition[left[0]]
+    partition[partition == partition[right[1]]] = partition[right[0]]
+
+#------------------------------------------------------------------------------#
+
+folder = '/home/elle/ANIGMA/porepy/examples/exampleXX/'
```
Comment (keileg) [inline examples/example9/continuous.py]: This is not going to work.

File: examples/example9/discontinuous.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    left = seeds[g.cell_centers[0, seeds] < 0.5]
+    right = seeds[g.cell_centers[0, seeds] > 0.5]
+
+    partition[partition == partition[left[1]]] = partition[left[0]]
+    partition[partition == partition[right[1]]] = partition[right[0]]
+
+#------------------------------------------------------------------------------#
+
+folder = '/home/elle/ANIGMA/porepy/examples/exampleXX/'
```
Comment (keileg) [inline examples/example9/discontinuous.py]: Local folder, absolute address.

File: src/porepy/fracs/importer.py [hunk 3]
```diff
@@ -24,16 +24,18 @@ def from_csv(f_name, mesh_kwargs, domain=None, pause=False, **kwargs):
...
     pts, edges = fractures_from_csv(f_name, **kwargs)
+    f_set = np.array([pts[:, e] for e in edges.T])
 
     # Define the domain as bounding-box if not defined
     if domain is None:
         overlap = kwargs.get('domain_overlap', 0)
         domain = _bounding_box(pts, overlap)
-
-    f_set = np.array([pts[:, e] for e in edges.T])
+        return meshing.simplex_grid(f_set, domain, **mesh_kwargs), domain
```
Comment (keileg) [inline src/porepy/fracs/importer.py]: Not sure about the context here, but it seems lines 38 and 40 returns a different number of parameters. Is this on purpose, and is it wise?

File: src/porepy/fracs/meshing.py [hunk 4]
```diff
@@ -84,6 +84,55 @@ def simplex_grid(fracs, domain, **kwargs):
     split_grid.split_fractures(gb)
     return gb
 
+#------------------------------------------------------------------------------#
+
+def from_gmsh(file_name, dim, **kwargs):
+    """
+    Import an already generated grid from gmsh.
+    NOTE: Only 2d grid is implemented so far.
```
Comment (keileg) [inline src/porepy/fracs/meshing.py]: Please add a comment on what is missing to make 3d work.

File: src/porepy/grids/gmsh/mesh_2_grid.py [hunk 5]
```diff
@@ -178,10 +178,11 @@ def create_0d_grids(pts, cells):
 
 
 def create_embedded_line_grid(loc_coord, glob_id):
+    atol = 1e-4
```
Comment (keileg) [inline src/porepy/grids/gmsh/mesh_2_grid.py]: Should it be possible to set this externally?

File: src/porepy/utils/comp_geom.py [hunk 6]
```diff
@@ -1594,9 +1595,14 @@ def is_collinear(pts, tol=1e-5):
...
+    dist = 0
+    for i in np.arange(pts.shape[1]):
+        for j in np.arange(i+1, pts.shape[1]):
+            dist = max(dist, np.linalg.norm(pts[:, i] - pts[:, j]))
+
     coll = np.array([np.linalg.norm(np.cross(p - pt0, pt1 - pt0)) \
-             for p in pts[:, 1:-1].T])
-    return np.allclose(coll, np.zeros(coll.size), rtol=tol)
+             for p in pts[:, 1:-1].T])/dist
+    return np.allclose(coll, np.zeros(coll.size), atol=tol, rtol=0)
```
Comment (keileg) [inline src/porepy/utils/comp_geom.py]: atol, and not rtol? I'm not sure.

File: src/porepy/utils/comp_geom.py [hunk 7]
```diff
@@ -1741,6 +1747,48 @@ def distance_segment_segment(s1_start, s1_end, s2_start, s2_end):
...
+    Compute the minimum distance between a point and a segment.
+
+    Parameters:
+        pt: the point
+        start: a point representing one extreme of the segment.
+        end: the second point representing the segment.
+    Returns:
+        distance: the minimum distance between the point and the segment.
+        intersect: point of intersection
+        if_pt: True if the point on intersection is one of the input points (pt,
```
Comment (keileg) [inline src/porepy/utils/comp_geom.py]: Do you really return if_pt? If so, what's the tolerance?

File: src/porepy/utils/comp_geom.py [hunk 8]
```diff
@@ -1741,6 +1747,48 @@ def distance_segment_segment(s1_start, s1_end, s2_start, s2_end):
     dist = d_starts + sc * d1 - tc * d2
     return np.sqrt(dist.dot(dist))
 
+#------------------------------------------------------------------------------#
+
+def distance_point_segment(pt, start, end):
```
Comment (keileg) [inline src/porepy/utils/comp_geom.py]: I have implemented the same function, although allowing for multiple points and segments, see branch frac_mesh_3d. I suggest leaving this for now, but replacing it with the more general function when the latter branch is merged.

File: test/integration/test_coarsening.py [hunk 9]
```diff
@@ -39,9 +42,10 @@ def test_coarse_grid_2d( self ):
 
 #------------------------------------------------------------------------------#
 
-    def test_coarse_grid_3d( self ):
+    def test_coarse_grid_3d(self):
         g = structured.CartGrid([2, 2, 2])
-        g = generate_coarse_grid( g, [0, 0, 0, 0, 1, 1, 2, 2] )
+        g.compute_geometry()
+        co.generate_coarse_grid( g, [0, 0, 0, 0, 1, 1, 2, 2] )
```
Comment (keileg) [inline test/integration/test_coarsening.py]: Pylint is not going to like the white spaces in ( g ). Applies to a lot of places.

File: examples/example9/continuous.py (hunk 1 above)
Comment (alessiofumagalli) [inline examples/example9/continuous.py]: you're right. I just add this example to be on the safe side, it's not really in it's final version but almost. The same for your next comment.

File: src/porepy/fracs/importer.py (hunk 3 above)
Comment (alessiofumagalli) [inline src/porepy/fracs/importer.py]: Well, if don't know the domain (bounding-box) a-priori we can let the code computes it. That's my idea.

File: src/porepy/utils/comp_geom.py (hunk 8 above)
Comment (alessiofumagalli) [inline src/porepy/utils/comp_geom.py]: ok

File: test/integration/test_coarsening.py (hunk 9 above)
Comment (alessiofumagalli) [inline test/integration/test_coarsening.py]: ok, I'll try to fix it..

File: src/porepy/utils/comp_geom.py (hunk 7 above)
Comment (alessiofumagalli) [inline src/porepy/utils/comp_geom.py]: OK, if_pt was referred to an old implementation.

File: src/porepy/utils/comp_geom.py (hunk 6 above)
Comment (alessiofumagalli) [inline src/porepy/utils/comp_geom.py]: not sure as well.. tolerance will kill us in many terrible ways :(

File: src/porepy/grids/gmsh/mesh_2_grid.py (hunk 5 above)
Comment (alessiofumagalli) [inline src/porepy/grids/gmsh/mesh_2_grid.py]: moved as input with default value

**Pull request merged**